File: das_client/formation/lib/src/model/formation.dart
import 'package:collection/collection.dart';
import 'package:formation/src/model/formation_run.dart';

class Formation({
  required final String operationalTrainNumber,
  required final String company,
  required final DateTime operationalDay,
  final List<FormationRun> formationRuns = const [],
}) {
  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Formation &&
          runtimeType == other.runtimeType &&
          operationalTrainNumber == other.operationalTrainNumber &&
          company == other.company &&
          operationalDay == other.operationalDay &&
          ListEquality().equals(formationRuns, other.formationRuns);

  @override
  int get hashCode => Object.hash(operationalTrainNumber, company, operationalDay, formationRuns);
}

File: das_backend/src/main/java/ch/sbb/das/backend/cargo/api/v1/model/Formation.java
package ch.sbb.das.backend.cargo.api.v1.model;

import static ch.sbb.das.backend.cargo.api.v1.FormationController.OPERATIONAL_DAY_DESCRIPTION;
import static ch.sbb.das.backend.cargo.api.v1.FormationController.OPERATIONAL_TRAIN_NUMBER_DESCRIPTION;

import ch.sbb.das.backend.companies.CompanyCode;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Schema;
import java.time.LocalDate;
import java.util.List;
import lombok.Builder;

@Builder
@Schema(description = "The formation represents a composition of implied wagons of a train journey (de: Zuglauf).")
public record Formation(

    @Schema(description = OPERATIONAL_TRAIN_NUMBER_DESCRIPTION, requiredMode = Schema.RequiredMode.REQUIRED)
    String operationalTrainNumber,

    @Schema(description = OPERATIONAL_DAY_DESCRIPTION, requiredMode = Schema.RequiredMode.REQUIRED)
    LocalDate operationalDay,

    @Schema(description = CompanyCode.DESCRIPTION, requiredMode = Schema.RequiredMode.REQUIRED)
    CompanyCode company,

    @ArraySchema(arraySchema = @Schema(requiredMode = Schema.RequiredMode.REQUIRED, description = "The order of items match the actual run order."), minItems = 1)
    List<FormationRun> formationRuns
) {

}
